File: SnowForecast/app/src/main/java/com/sners/snowforecast/data/CurrentlyData.java
package com.sners.snowforecast.data;

import java.text.DecimalFormat;
import java.util.ArrayList;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import com.sners.snowforecast.data.*;


/*
This class is a convenience summary.
The data it uses comes from minutely, but it calculates things once
 */
public class CurrentlyData   {

	private String headline = "Not found";
	private String icon = "";
	private float precipIntensity = 0;
	private float precipProbability = 0;
	private Integer weatherCode;
	private float temperature = 0;
	private float tempFeelsLike = 0;
	private float windSpeed = 0;	//	Wind speed is metres per second in the data classes
	private float nextRain = 0;
	private float nextSnow = 0;
	private String time;
	private String sunsetTime;
	private String sunriseTime;
	private ArrayList<String> galleryIcons;
	private ArrayList<String> weatherWords;


	/*
	Constructor
	 */
	public CurrentlyData(JSONObject currentJson)
	{
		try {
			headline = currentJson.getString(WeatherConstants.CONDITIONS);
			icon = currentJson.getString(WeatherConstants.ICON);
			time = currentJson.getString(WeatherConstants.TIME);
			sunriseTime = currentJson.getString(WeatherConstants.SUNRISE_TIME);
			sunsetTime = currentJson.getString(WeatherConstants.SUNSET_TIME);
			precipIntensity = (float) currentJson.getDouble(WeatherConstants.PRECIP_INTENSITY);
			temperature = (float) currentJson.getDouble(WeatherConstants.TEMPERATURE);
			tempFeelsLike = (float) currentJson.getDouble(WeatherConstants.TEMP_FEELS_LIKE);
			if (precipIntensity > 0) {
				//	This is null if no precipIntensity
				precipProbability = (float) currentJson.getDouble(WeatherConstants.PRECIP_PROBABILITY);
			}

			if (precipIntensity > 0 && precipProbability > 0) {
				JSONArray weatherWordsJson = currentJson.getJSONArray(WeatherConstants.PRECIP_TYPE);
				for (int wordCounter = 0 ; wordCounter < weatherWordsJson.length() ; wordCounter++) {
					String precipType = weatherWordsJson.getString(wordCounter);
					weatherWords.add(precipType) ;
				}
			}

			windSpeed = (float) currentJson.getDouble(WeatherConstants.WIND_SPEED);
		} catch (JSONException e) {
			e.printStackTrace();
		}




		galleryIcons = new ArrayList<String>();
	}


	public String getHeadline() {
		return headline;
	}

	public void setHeadline(String headline) {
		this.headline = headline;
	}

	public String getIcon() {
		return icon;
	}

	public void setIcon(String icon) {
		this.icon = icon;
	}

	public  float getTemperature() {
		return temperature;
	}

	public void setTemperature(float temperature) {
		this.temperature = temperature;
	}

	public float getPrecipIntensity() {
		return precipIntensity;
	}

	public float getPrecipProbability() {
		return precipProbability;
	}

	public float getWindSpeed() {
		return windSpeed;
	}

	public ArrayList<String> getGalleryIcons() {
		return galleryIcons;
	}

	public void setGalleryIcons(ArrayList<String> galleryIcons) {
		this.galleryIcons = galleryIcons;
	}

	public String getTime() {
		return time;
	}

		public String getSunsetTime() {
		return sunsetTime;
	}

	public String getSunriseTime() {
		return sunriseTime;
	}

	public ArrayList<String> getWeatherWords() {
		return weatherWords;
	}

	public float getTempFeelsLike() {
		return tempFeelsLike;
	}
}

File: SnowForecast/app/src/main/java/com/sners/snowforecast/data/CurrentlyData.kt
package com.sners.snowforecast.data

import com.sners.snowforecast.weather.WeatherConstants
import org.json.JSONException
import org.json.JSONObject

/**
 * A data class of current conditions
 * It takes the raw json and converts it into a data structure
 * @param currentJson The raw current data coming in from the API call
 */
class CurrentlyData(currentJson: JSONObject) : IntervalData() {
    /**
     *  @property headline Top level summary
     */
    var headline: String = "Not found"
        private set

    /**
     *  @property icon Name of the icon that represents current conditions
     */
    var icon: String = ""
        private set


    /**
     *  @property sunsetTime Sunset time as read from the json
     */
    private var sunsetTime: String? = null
        private set

    /**
     *  @property sunriseTime Sunrise time as read from the json
     */
    private var sunriseTime: String? = null
        private set

    /*
  Constructor
   */
    init {
        try {
            /* For each property, unpack the Json string as appropriate */
            this.headline = currentJson.getString(WeatherConstants.CONDITIONS)
            this.icon = currentJson.getString(WeatherConstants.ICON)
            //  Times
            time = currentJson.getString(WeatherConstants.TIME)
            sunriseTime = currentJson.getString(WeatherConstants.SUNRISE_TIME)
            sunsetTime = currentJson.getString(WeatherConstants.SUNSET_TIME)
            //  Temperature
            temperature = readTemperatureFrom(currentJson)
            tempFeelsLike = currentJson.getDouble(WeatherConstants.TEMP_FEELS_LIKE).toFloat()
            //  Precipitation
            this.precipIntensity = currentJson.getDouble(WeatherConstants.PRECIP_INTENSITY).toFloat()
            cloudCover = readCloudCoverFrom(currentJson)

            if (this.precipIntensity > 0) {
                //	This is null if no precipIntensity
                this.precipProbability = currentJson.getDouble(WeatherConstants.PRECIP_PROBABILITY).toFloat()
            }
            if (this.precipIntensity > 0 && this.precipProbability > 0) {
                val weatherWordsJson = currentJson.getJSONArray(WeatherConstants.PRECIP_TYPE)
                for (wordCounter in 0..weatherWordsJson.length() - 1) {
                    val precipType = weatherWordsJson.getString(wordCounter)
                    weatherWords.add(precipType)
                }
            }
            //  Wind
            this.windSpeed = currentJson.getDouble(WeatherConstants.WIND_SPEED)
            this.windGusts = currentJson.getDouble(WeatherConstants.WIND_GUST).toFloat()
            val dir = currentJson.getString(WeatherConstants.WIND_DIRECTION)
            //  If winddir is not entered, it should not be initialised
            if (dir.isNotEmpty()) {
                windDir = dir.toFloat()
            }        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    /**
     * Getters
     */
    /**
     *  Get Sunrise Time
     *  @return sunrise time as a string or empty string if not present
     */
    fun getSunriseTime() : String {
        return if (null == this.sunriseTime) {
            ""
        } else {
            sunriseTime as String
        }
    }

    /**
     *  Get Sunset Time
     *  @return sunset time as a string or empty string if not present
     */
    fun getSunsetTime() : String {
        return if (null == this.sunsetTime) {
            ""
        } else {
            sunsetTime as String
        }
    }
}

